app/controllers/income/estimatedPay/update/IncomeUpdateCalculatorController.scala

```diff
@@ -72,49 +74,111 @@ class IncomeUpdateCalculatorController @Inject()(incomeService: IncomeService,
   with FormValuesConstants
   with FeatureTogglesConfig {
 
-  def estimatedPayLandingPage(id: Int): Action[AnyContent] = authorisedForTai(personService).async { implicit user =>
+  def onPageLoad(id: Int): Action[AnyContent] = authorisedForTai(personService).async { implicit user =>
     implicit person =>
       implicit request =>
 
-        val taxCodeIncomesFuture = taxAccountService.taxCodeIncomes(Nino(user.getNino), TaxYear())
         val employmentFuture = employmentService.employment(Nino(user.getNino), id)
+        val estimatedPayCompletionFuture = estimatedPayJourneyCompletionService.hasJourneyCompleted(id.toString)
 
         for {
-          taxCodeIncomeDetails <- taxCodeIncomesFuture
           employmentDetails <- employmentFuture
+          estimatedPayCompletion <- estimatedPayCompletionFuture
         } yield {
-          (taxCodeIncomeDetails, employmentDetails) match {
-            case (TaiSuccessResponseWithPayload(taxCodeIncomes: Seq[TaxCodeIncome]), Some(employment)) =>
-              val taxCodeIncomeSource = taxCodeIncomes.find(_.employmentId.contains(id)).
-                getOrElse(throw new RuntimeException(s"Income details not found for employment id $id"))
-              val isPension = taxCodeIncomeSource.componentType == PensionIncome
-              Ok(views.html.incomes.estimatedPayLandingPage(employment.name, id, isPension))
+          employmentDetails match {
+            case Some(employment) =>
+
+              val incomeType = incomeTypeIdentifier(employment.receivingOccupationalPension)
+
+              journeyCache(cacheMap = Map(
+                UpdateIncome_NameKey -> employment.name,
+                UpdateIncome_IdKey -> id.toString,
+                UpdateIncome_IncomeTypeKey -> incomeType))
+
```

Comment:
This should be completed before the redirect, the cache could still be running in a background thread, the redirect happens and we get a key not found error on the next page. So we should be mapping/for comprehension over these

Reply:
Fixed it now